packages/gatsby-source-graphql/src/batching/dataloader-link.ts

```diff
@@ -0,0 +1,81 @@
+import DataLoader from "dataloader"
+import { ApolloLink, Observable, Operation, FetchResult } from "apollo-link"
+import { print } from "graphql"
+import { IQuery, IQueryResult, merge, resolveResult } from "./merge-queries"
+
+interface IOptions {
+  uri: string
+  fetch: Function
+  fetchOptions?: object
+  dataLoaderOptions?: object
+  headers?: object
+}
+
+export function createDataloaderLink(options: IOptions): ApolloLink {
+  const load = async (keys: ReadonlyArray<IQuery>): Promise<IQueryResult[]> => {
+    const query = merge(keys)
+    const result: object = await request(query, options)
+    if (!isValidGraphQLResult(result)) {
+      const error = extractFirstError(result)
+      throw new Error(`Failed to load query batch: ${error}`)
+    }
+    return resolveResult(result)
+  }
+
+  const dataloader = new DataLoader(load, {
+    cache: false,
+    maxBatchSize: 5,
```

Comment:
I'm thinking that maybe we should default to current max concurrency as maxBatchSize. Thoughts?

Reply:
So the thing is - concurrency gives more performance benefits than batching. Making `maxBatchSize` equal to concurrency limit means that we send a single query sequentially - this works roughly the same as sending queries concurrently.

Batching starts shining when you hit the concurrency limits of your server. Another possible issue with batching is that huge batched queries may overload servers or just fail the complexity scoring.

So to recap we can do `maxBatchSize = concurrencyLimit / optimialConcurrencyLevel`

Unfortunately `optimialConcurrencyLevel` will be different for different servers. We can probably stay on a safer side and set it to 4 or something like that.

Reply:
Sounds good to me, let's do smth like that.